### PR TITLE
chore: add more vigilant checking to resolution

### LINKED ISSUE #100: In blocking functions like `subscribe`, ensure the event to resolve on occurred from the same topic
Based on https://github.com/WalletConnect/web3inbox/issues/416 which led to the discovery of this issue. 

The issue above occurs because going on w3i app and clicking through multiple dapps before there is time to resolve the notifications causes them to show up under the wrong dapp because they resolve under the wrong topic.

Eg:
1. **(User action)** Go to Dapp A notifications
2. **(Function call issued from web3inbox)**: `getNotificationHistory(topic: "<A-sub-topic>")`
3. **(User action)** Before they have time to resolve, almost instantly switch to Dapp B
4. **(Function call issued from web3inbox)**: `getNotificationHistory(topic: "<B-sub-topic>")`
5. **(User action)** Wait
6. **(Internal logic of `getNotificationHistory(topic: "<B-sub-topic>")` call)**: Receive event `notify_get_notifications_response` which is in reality a response on `A-sub-topic`, **not** `B-sub-topic`
7. **(Internal logic of `getNotificationHistory(topic: "<B-sub-topic>")` call)**: `resolve` with notifications of `A`
8. **Final result**: Call for notifications under `B-sub-topic` resolves with `A-sub-to-pic`

### PR DESCRIPTION
**Pull request by devceline**
- Resolves #100 
- Only resolve blocking functions when the same topic is sent in the response as the request
- Add unit tests for blocking functions

### CODE REVIEW

**Review by arein — APPROVED** [review]
Is this a hotfix and you'll write tests later or why are there no tests covering this?

**Comment by devceline** [comment]
> Is this a hotfix and you'll write tests later or why are there no tests covering this?

Still working on the PR :D forgot to mark as draft

File: packages/notify-client/src/controllers/engine.ts [hunk 1]
```diff
@@ -287,13 +292,22 @@ export class NotifyEngine extends INotifyEngine {
...
 
     return new Promise<boolean>((resolve) => {
-      this.client.once("notify_subscription", (args) => {
+      const listener = (
+        args: NotifyClientTypes.EventArguments["notify_subscription"]
+      ) => {
+        if (args.topic === subscribeTopic) {
+          return;
+        }
+        this.client.off("notify_subscription", listener);
```
Comment (enesozturk) [inline packages/notify-client/src/controllers/engine.ts:301]: @devceline Is this approach closing/removing any event listener to check if it already exists? I am only curious why we were required to call `.off`s.

File: packages/notify-client/test/clients.spec.ts [hunk 2]
```diff
@@ -844,6 +844,112 @@ describe("Notify", () => {
...
+        expect(wallet.subscriptions.length).toEqual(1);
+        expect(wallet.subscriptions.getAll()[0].metadata.appDomain).toEqual(
+          testDappMetadata.appDomain
+        );
+      });
+
+      it.skipIf(!hasTestProjectSecret)(
+        "Only resolves when topic accurate response is issued",
+        async () => {
+          const preparedRegistration = await wallet.prepareRegistration({
```
Comment (enesozturk) [inline packages/notify-client/test/clients.spec.ts:874]: Could this register logic be implemented in `beforeEach`?

```js
beforeEach(()=>{
  const preparedRegistration = await wallet.prepareRegistration({
    account,
    domain: testDappMetadata.appDomain,
    allApps: true,
  });

  await wallet.register({
    registerParams: preparedRegistration.registerParams,
    signature: await onSign(preparedRegistration.message),
  });

```

**Review by enesozturk — APPROVED** [review]
LGTM! Left a few comments

File: packages/notify-client/src/controllers/engine.ts (hunk 1 above)
Comment (devceline) [reply packages/notify-client/src/controllers/engine.ts:301]: This is actually a bug where it needs to be `!==` just like the others below. Basically if we don't get the response from the topic we want, we keep the listener open

File: packages/notify-client/test/clients.spec.ts (hunk 2 above)
Comment (devceline) [reply packages/notify-client/test/clients.spec.ts:874]: Some tests need to pass different params, I'm okay with the code repetition for now honestly, I like redundancy > complexity in unit tests you can easily change a small element of the unit test to see what updates.

**Pull request merged**